From 3f3d0258215fc518bc4dcfa61dcfa44b6250f2bb Mon Sep 17 00:00:00 2001
From: srinidhira0 <srinidhi.rao@broadcom.com>
Date: Wed, 3 Dec 2025 10:48:10 +0000
Subject: [PATCH] crypto:ecb: Set ecb alg flag as Internal.

  -  ECB algorithm is not part of canister and there is no requirement
     to expose ECB to external interfaces. However, ECB is still required
     by other AES variants such as ctr-aes, gcm-aes etc. Due to this
     fips_allowed flag for ecb cannot be disabled as dependent algo's
     self test will fail in FIPS mode.
     Hence set cra_flags for ECB algo as CRYPTO_ALG_FIPS_INTERNAL such that
     ECB is not exposed through AF_ALG but will be available other AES
     variants.

Signed-off-by: srinidhira0 <srinidhi.rao@broadcom.com>
---
 arch/x86/crypto/aesni-intel_glue.c |  2 +-
 crypto/ecb.c                       | 12 ++++++++++++
 crypto/skcipher.c                  |  8 ++++++++
 3 files changed, 21 insertions(+), 1 deletion(-)

diff --git a/arch/x86/crypto/aesni-intel_glue.c b/arch/x86/crypto/aesni-intel_glue.c
index cbed67a0d782..43b5d3aa3c2d 100644
--- a/arch/x86/crypto/aesni-intel_glue.c
+++ b/arch/x86/crypto/aesni-intel_glue.c
@@ -695,7 +695,7 @@ static struct skcipher_alg aesni_skciphers[] = {
 			.cra_name		= "__ecb(aes)",
 			.cra_driver_name	= "__ecb-aes-aesni",
 			.cra_priority		= 400,
-			.cra_flags		= CRYPTO_ALG_INTERNAL,
+			.cra_flags		= CRYPTO_ALG_INTERNAL | CRYPTO_ALG_FIPS_INTERNAL,
 			.cra_blocksize		= AES_BLOCK_SIZE,
 			.cra_ctxsize		= CRYPTO_AES_CTX_SIZE,
 			.cra_module		= THIS_MODULE,
diff --git a/crypto/ecb.c b/crypto/ecb.c
index e3a67789050e..b9fb7a995b5e 100644
--- a/crypto/ecb.c
+++ b/crypto/ecb.c
@@ -12,6 +12,7 @@
 #include <linux/kernel.h>
 #include <linux/module.h>
 #include <linux/slab.h>
+#include <linux/fips.h>
 
 static int crypto_ecb_crypt(struct crypto_cipher *cipher, const u8 *src,
 			    u8 *dst, unsigned nbytes, bool final,
@@ -126,6 +127,14 @@ static struct lskcipher_instance *lskcipher_alloc_instance_simple2(
 	inst->free = lskcipher_free_instance_simple2;
 
 	/* Default algorithm properties, can be overridden */
+	/*
+	 * Mark ECB as Internal as it is outside FIPS canister and
+	 * there is no requirement for the ECB to be exposed to user space
+	 * through AF_ALG
+	 */
+	if (fips_enabled)
+		inst->alg.co.base.cra_flags |= CRYPTO_ALG_FIPS_INTERNAL;
+
 	inst->alg.co.base.cra_blocksize = cipher_alg->cra_blocksize;
 	inst->alg.co.base.cra_alignmask = cipher_alg->cra_alignmask;
 	inst->alg.co.base.cra_priority = cipher_alg->cra_priority;
@@ -189,6 +198,9 @@ static int crypto_ecb_create(struct crypto_template *tmpl, struct rtattr **tb)
 	if (cipher_alg->co.ivsize)
 		return -EINVAL;
 
+	if (fips_enabled)
+		inst->alg.co.base.cra_flags |= CRYPTO_ALG_FIPS_INTERNAL;
+
 	inst->alg.co.base.cra_ctxsize = cipher_alg->co.base.cra_ctxsize;
 	inst->alg.setkey = cipher_alg->setkey;
 	inst->alg.encrypt = cipher_alg->encrypt;
diff --git a/crypto/skcipher.c b/crypto/skcipher.c
index fd3273b519dc..492168f2f11f 100644
--- a/crypto/skcipher.c
+++ b/crypto/skcipher.c
@@ -23,6 +23,8 @@
 #include <linux/seq_file.h>
 #include <linux/slab.h>
 #include <linux/string.h>
+#include <linux/fips.h>
+#include <linux/crypto.h>
 #include <net/netlink.h>
 #include "skcipher.h"
 
@@ -917,6 +919,12 @@ int crypto_register_skcipher(struct skcipher_alg *alg)
 	struct crypto_alg *base = &alg->base;
 	int err;
 
+	/* ECB is not exposed through AF_ALG when fips_enabled is set. */
+	if (unlikely(!fips_enabled &&
+			strnstr(base->cra_driver_name, "ecb-aes-aesni",
+			CRYPTO_MAX_ALG_NAME)))
+		base->cra_flags &= ~CRYPTO_ALG_FIPS_INTERNAL;
+
 	err = skcipher_prepare_alg(alg);
 	if (err)
 		return err;
-- 
2.35.6

